### PR TITLE
Propagate reattach info

### PR DESCRIPTION
**Pull request by kmoe**
Requires https://github.com/hashicorp/terraform-exec/pull/78

This PR fixes issues introduced in #35, in which Terraform commands were reimplemented using terraform-exec. The tfexec paradigm does not permit transparent propagation of environment variables through to the Terraform CLI process, which meant that the crucial `TF_PROVIDERS_REATTACH` env var was not present in the binary testing environment. This behaviour, and the ability for the caller to set other env vars at arbitrary times, is restored by allowing `WorkingDir` to maintain a pseudo-environment.

### Add `wd.SetReattachInfo(reattachInfo string)` and `wd.UnsetReattachInfo()`

The caller provides the reattach info (the contents of `TF_PROVIDERS_REATTACH`), and this is stored on `WorkingDir` and added to each tfexec command as necessary.

### ~Restore previous behaviour of `wd.Setenv(envVar, val string)`~ (reverted, see comment below)

~Since terraform-plugin-test is no longer in control of the environment variables for each Terraform command, we instead ensure the correct environment is set before each `tfexec.Terraform` command.~

~Note: `wd.Setenv` and `wd.Unsetenv` are not used in the SDK, and should be removed if there is no clear use case. This is a breaking change and so will be dealt with separately as https://github.com/hashicorp/terraform-plugin-test/issues/39.~

### CODE REVIEW

File: working_dir.go [hunk 1]
```diff
@@ -94,6 +119,15 @@ func (wd *WorkingDir) SetConfig(cfg string) error {
...
 	}
 
+	tf.SetDisablePluginTls(true)
+	tf.SetSkipProviderVerify(true)
+
+	if p := os.Getenv("TF_ACC_LOG_PATH"); p != "" {
+		tf.SetLogPath(p)
+	}
+
+	tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: There's an error here we're not checking.

File: working_dir.go [hunk 2]
```diff
@@ -163,7 +197,8 @@ func (wd *WorkingDir) Init() error {
 		return fmt.Errorf("must call SetConfig before Init")
 	}
 
-	return wd.tf.Init(context.Background(), tfexec.Dir(wd.configDir))
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 3]
```diff
@@ -183,7 +218,9 @@ func (wd *WorkingDir) planFilename() string {
 // CreatePlan runs "terraform plan" to create a saved plan file, which if successful
 // will then be used for the next call to Apply.
 func (wd *WorkingDir) CreatePlan() error {
-	return wd.tf.Plan(context.Background(), tfexec.Refresh(false), tfexec.Out("tfplan"), tfexec.Dir(wd.configDir))
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 4]
```diff
@@ -199,18 +236,17 @@ func (wd *WorkingDir) RequireCreatePlan(t TestControl) {
 // CreateDestroyPlan runs "terraform plan -destroy" to create a saved plan
 // file, which if successful will then be used for the next call to Apply.
 func (wd *WorkingDir) CreateDestroyPlan() error {
-	args := []string{"plan", "-destroy", "-refresh=false"}
-	args = append(args, wd.baseArgs...)
-	args = append(args, "-out=tfplan", wd.configDir)
-	return wd.runTerraform(nil, args...)
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked

File: working_dir.go [hunk 5]
```diff
@@ -224,6 +260,7 @@ func (wd *WorkingDir) Apply() error {
 		}
 		args = append(args, tfexec.DirOrPlan(configDir))
 	}
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 6]
```diff
@@ -243,7 +280,8 @@ func (wd *WorkingDir) RequireApply(t TestControl) {
 // If destroy fails then remote objects might still exist, and continue to
 // exist after a particular test is concluded.
 func (wd *WorkingDir) Destroy() error {
-	return wd.tf.Destroy(context.Background(), tfexec.Refresh(false), tfexec.Dir(wd.configDir))
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 7]
```diff
@@ -276,7 +314,8 @@ func (wd *WorkingDir) SavedPlan() (*tfjson.Plan, error) {
 		return nil, fmt.Errorf("there is no current saved plan")
 	}
 
-	return wd.tf.ShowPlanFile(context.Background(), wd.planFilename())
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 8]
```diff
@@ -302,9 +341,10 @@ func (wd *WorkingDir) SavedPlanStdout() (string, error) {
 
 	var ret bytes.Buffer
 
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 9]
```diff
@@ -328,7 +368,8 @@ func (wd *WorkingDir) RequireSavedPlanStdout(t TestControl) string {
 //
 // If the state cannot be read, State returns an error.
 func (wd *WorkingDir) State() (*tfjson.State, error) {
-	return wd.tf.Show(context.Background())
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 10]
```diff
@@ -345,7 +386,8 @@ func (wd *WorkingDir) RequireState(t TestControl) *tfjson.State {
 
 // Import runs terraform import
 func (wd *WorkingDir) Import(resource, id string) error {
-	return wd.tf.Import(context.Background(), resource, id, tfexec.Config(wd.configDir))
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 11]
```diff
@@ -360,7 +402,8 @@ func (wd *WorkingDir) RequireImport(t TestControl, resource, id string) {
 
 // Refresh runs terraform refresh
 func (wd *WorkingDir) Refresh() error {
-	return wd.tf.Refresh(context.Background(), tfexec.State(filepath.Join(wd.baseDir, "terraform.tfstate")), tfexec.Dir(wd.configDir))
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

File: working_dir.go [hunk 12]
```diff
@@ -377,6 +420,7 @@ func (wd *WorkingDir) RequireRefresh(t TestControl) {
 //
 // If the schemas cannot be read, Schemas returns an error.
 func (wd *WorkingDir) Schemas() (*tfjson.ProviderSchemas, error) {
+	wd.tf.SetEnv(wd.buildEnv())
```
Comment (paddycarver) [inline working_dir.go]: Error here should be checked.

**Review by paddycarver — CHANGES_REQUESTED** [review]
Called out some errors that need checked (sorry for the repetition).

We could also automate this with https://github.com/kisielk/errcheck or golangci-lint.

**Comment by kmoe** [comment]
After some discussion, I'm removing the propagation of the whole `os.Environ()` into the tfexec environment, because it causes unexpected results with reattach tests. Leaving the two (now a little counterintuitive, and in any case unused) `wd.Setenv()` and `wd.Unsetenv()`, which will be removed when tftest is very shortly moved into the SDK itself.

**Review by paddycarver — APPROVED** [review]

**Pull request merged**